libcontainer/cgroups/fs/freezer.go

```diff
@@ -28,33 +28,49 @@ func (s *FreezerGroup) Apply(path string, d *cgroupData) error {
 
 func (s *FreezerGroup) Set(path string, cgroup *configs.Cgroup) error {
 	switch cgroup.Resources.Freezer {
-	case configs.Frozen, configs.Thawed:
-		for {
-			// In case this loop does not exit because it doesn't get the expected
-			// state, let's write again this state, hoping it's going to be properly
-			// set this time. Otherwise, this loop could run infinitely, waiting for
-			// a state change that would never happen.
-			if err := fscommon.WriteFile(path, "freezer.state", string(cgroup.Resources.Freezer)); err != nil {
+	case configs.Frozen:
+		// As per kernel doc (freezer-subsystem.txt), if FREEZING
+		// is seen, userspace should either retry or thaw.
+		//
+		// Apparently, kernel can't reliably freeze a cgroup
+		// while new processes are appearing in it (either via
+		// fork() / clone() or by writing new PIDs to the cgroup).
+		//
+		// The particulars in the code below (number of retries
+		// and no sleep in between) are chosen empirically to
+		// minimize the number of failured freezes.
+		for i := 0; i < 1000; i++ {
+			if err := fscommon.WriteFile(path, "freezer.state", string(configs.Frozen)); err != nil {
 				return err
 			}
 
-			state, err := s.GetState(path)
+			state, err := fscommon.ReadFile(path, "freezer.state")
 			if err != nil {
 				return err
 			}
-			if state == cgroup.Resources.Freezer {
-				break
+			state = strings.TrimSpace(state)
+			switch state {
+			case "FREEZING":
+				continue
+			case string(configs.Frozen):
+				return nil
+			default:
+				// should never happen
+				return fmt.Errorf("unexpected state %s while freezing", strings.TrimSpace(state))
 			}
-
-			time.Sleep(1 * time.Millisecond)
```

Comment:
Just to make sure I understand why this was dropped -- is the idea that we're not waiting on any other process to finish so waiting just inflates `runc pause` times?

Reply:
In essence, yes.

I tried with different values of Sleep (small, large, ever-increasing, ever-decreasing, random) and with different reproducers (exec vs pause/resume, exec vs update, fork vs pause/resume, pthread bomb vs pause/resume) and the only effect I saw from sleeping is the increased time to freeze. In other words, I was not able to see that sleep improves the chances of freezing, so I dropped it entirely.

Reply:
Makes sense.